cmake/hipsycl-config.cmake.in

```diff
@@ -48,7 +51,14 @@ if(HIPSYCL_ROCM_BACKEND_AVAILABLE)
   endif()
 endif()
 
-set(HIPSYCL_PLATFORM "${HIPSYCL_PLATFORM}" CACHE STRING "The platform that hipSYCL should target. One of ${HIPSYCL_PLATFORMS_STRING}.")
+set(HIPSYCL_PLATFORM "" CACHE STRING
+    "The platform that hipSYCL should target. One of ${HIPSYCL_PLATFORMS_STRING}.")
+
+string(REPLACE "|" ";" HIPSYCL_PLATFORM_OPTIONS ${HIPSYCL_PLATFORMS_STRING})
+# Add canonical platform aliases to platform options list
+list(APPEND HIPSYCL_PLATFORM_OPTIONS "host" "hipcpu" "nvidia" "amd" "hip" "hcc")
```

Comment:
This list includes all potential options, even if those backends aren't available (as opposed to `HIPSYCL_PLATFORMS_STRING`, which only includes available backends). Besides, I don't think we should be advertising the non-canonical names, these exist mostly for backwards compatibility (and should be removed at some point imo).

That being said, I realize now that the whole mechanism of building ` HIPSYCL_PLATFORMS_STRING` is needlessly convoluted, as a simple list append followed by a `list(JOIN ...)`  would achieve the same thing. While you're at it, maybe you could change that..? :)

Reply:
> This list includes all potential options, even if those backends aren't available (as opposed to HIPSYCL_PLATFORMS_STRING, which only includes available backends). Besides, I don't think we should be advertising the non-canonical names, these exist mostly for backwards compatibility (and should be removed at some point imo).

I have modified the `HIPSYCL_PLATFORM` to accept values from a given list. I wasn't sure if the values from [lines 70-78][1] are required or not. Given that they are essentially aliases to the three options `cuda`, `rocm` and `cpu`, in case backwards compatibility is the reason they are present, I added them to the options list as well. Note that the default value of this variable is still "" empty so that environment variables can be used.

I can revert the change to `HIPSYCL_PLATFORM` to being a regular string where the user has to type in the value if that is preferred over the cycling through a list. Please let me know.

About the removing the aliases at some point, those options can be easily removed by deleting the below line.

```cmake
# Add canonical platform aliases to platform options list
list(APPEND HIPSYCL_PLATFORM_OPTIONS "host" "hipcpu" "nvidia" "amd" "hip" "hcc")
```

> That being said, I realize now that the whole mechanism of building HIPSYCL_PLATFORMS_STRING is needlessly convoluted, as a simple list append followed by a list(JOIN ...) would achieve the same thing. While you're at it, maybe you could change that..? :)

I agree too. The reason I left it out is the following:
1. Right now, `HIPSYCL_PLATFORMS_STRING` is ending up being something like ` cpu | cuda | rocm ` which is nice to read in doc string for that particular CMake option instead of  semi-colon separated list.
2. `HIPSYCL_PLATFORMS_STRING` can easily be converted to a CMake option/string list which can be used as option list for `HIPSYCL_PLATFORM`.

[1]: https://github.com/illuhad/hipSYCL/blob/develop/cmake/hipsycl-config.cmake.in#L70-L78

Reply:
> I can revert the change to HIPSYCL_PLATFORM to being a regular string where the user has to type in the value if that is preferred over the cycling through a list. Please let me know.

I don't use the CMake GUI, but it is my understanding that the `STRINGS` property on cache entries is merely a visual aid for the GUI, and doesn't perform any form of validation. I think it should thus (1) not offer the option of selecting platforms that aren't available, and (2) not show any aliases, giving the impression that there is any difference between them, if in fact there is not (and we don't want to encourage their usage in the first place). Bottom line is, I think showing the possible options in the GUI is a good UX improvement, only it should be the right set of options :).

> I agree too. The reason I left it out is the following:

Yes, what I meant is that we could go about it in the opposite direction: Instead of building that pipe-separated string with the cumbersome `if() / else()` logic, first the `HIPSYCL_PLATFORM_OPTIONS` list could be built ([here](https://github.com/illuhad/hipSYCL/blob/4aea57afbd37faa241b0dfe67088e66fa11b55de/cmake/hipsycl-config.cmake.in#L23), [here](https://github.com/illuhad/hipSYCL/blob/4aea57afbd37faa241b0dfe67088e66fa11b55de/cmake/hipsycl-config.cmake.in#L34) and [here](https://github.com/illuhad/hipSYCL/blob/4aea57afbd37faa241b0dfe67088e66fa11b55de/cmake/hipsycl-config.cmake.in#L45)) and then converted into the nice looking string using `list(JOIN HIPSYCL_PLATFORM_OPTIONS " | " HIPSYCL_PLATFORMS_STRING)`.

Reply:
> Point 1

It is not cmake GUI alone. CMake curses GUI i.e. the command line will also use this list of STRINGS. It doesn't do any validation, however it does help users using CMake curses GUI to choose the right option without the need to type anything. Okay, only show `cpu | cuda | rocm`, if the user needs to, they will enter the other values via environment variable. Shall I proceed for that change ?

> Point 2

Sure, I can look into changing that logic.

Reply:
> Okay, only show cpu | cuda | rocm, if the user needs to, they will enter the other values via environment variable. Shall I proceed for that change ?

Only show whatever is available, i.e., the contents of `HIPSYCL_PLATFORMS_STRING` (or rather, `HIPSYCL_PLATFORM_OPTIONS`, if you do that refactoring). They will never need to provide any other value.

> Sure, I can look into changing that logic.

Great, thank you!